### PR TITLE
Proposed changes for IDAM

### PR DESCRIPTION
**Pull request by nikola-naydenov-hmcts**
This is a pull request containing all the custom changes made in the branch "SIDM_increase_FuncTest_timout" currently used by the idam team to build and test web-apps.

Recently we have been falling behind with maintaining the branch with regards to changes made to master and the proposal to once more try and permanently merge the required changes into master and switch to using that instead was made.

I have created this PR so the reasoning behind the required changes, possible alternatives and a way forward can be discussed.

A short summary of why the changes are there is, the idam project has had an extra **_idam-_** prefix in it's resources, for honestly reasons not too clear to me, since it's inception.
Also the idam team cannot deploy continuously to AAT so a deployment in the Preview environment is where the majority of upcoming changes are being deployed and tested. Preview must be a **real** environment so to speak, not just for pull requests.

### CODE REVIEW

File: vars/sectionBuildAndTest.groovy [hunk 1]
```diff
@@ -29,14 +29,14 @@ def call(params) {
       withAcrClient(subscription) {
         projectBranch = new ProjectBranch(env.BRANCH_NAME)
         imageRegistry = env.TEAM_CONTAINER_REGISTRY ?: env.REGISTRY_NAME
-        acr = new Acr(this, subscription, imageRegistry, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION)
+        acr = new Acr(this, subscription, env.REGISTRY_NAME, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION, projectBranch)
```
Comment (timja) [inline vars/sectionBuildAndTest.groovy:32]: this looks like a merge conflict issue

```suggestion
        acr = new Acr(this, subscription, imageRegistry, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION, projectBranch)
```

File: vars/sectionDeployToAKS.groovy [hunk 2]
```diff
@@ -33,13 +33,14 @@ def call(params) {
...
   def imageRegistry
+  def projectBranch = new ProjectBranch(env.BRANCH_NAME)
 
   Builder builder = pipelineType.builder
 
   withAcrClient(subscription) {
     imageRegistry = env.TEAM_CONTAINER_REGISTRY ?: env.REGISTRY_NAME
-    acr = new Acr(this, subscription, imageRegistry, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION)
-    dockerImage = new DockerImage(product, component, acr, new ProjectBranch(env.BRANCH_NAME).imageTag(), env.GIT_COMMIT)
+    acr = new Acr(this, subscription, env.REGISTRY_NAME, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION, projectBranch)
```
Comment (timja) [inline vars/sectionDeployToAKS.groovy:42]:
```suggestion
    acr = new Acr(this, subscription, imageRegistry, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION, projectBranch)
```

File: vars/sectionDeployToAKS.groovy [hunk 3]
```diff
@@ -53,6 +54,8 @@ def call(params) {
             onPR {
               deploymentNumber = githubCreateDeployment()
             }
+            params.environment = params.environment.replace('idam-', '')
```
Comment (timja) [inline vars/sectionDeployToAKS.groovy:57]:
```suggestion
            params.environment = params.environment.replace('idam-', '') // hack to workaround incorrect idam environment value
```

File: vars/sectionDeployToAKS.groovy [hunk 4]
```diff
@@ -70,9 +73,11 @@ def call(params) {
         withTeamSecrets(config, environment) {
           stageWithAgent("Smoke Test - AKS ${environment}", product) {
             testEnv(aksUrl) {
-              pcr.callAround("smoketest:${environment}") {
-                timeoutWithMsg(time: 10, unit: 'MINUTES', action: 'Smoke Test - AKS') {
-                  builder.smokeTest()
+              pcr.callAround("smoketest-aks:${environment}") {
```
Comment (timja) [inline vars/sectionDeployToAKS.groovy:76]: why do you need a -aks version of this?

File: vars/sectionDeployToAKS.groovy [hunk 5]
```diff
@@ -81,9 +86,11 @@ def call(params) {
           onFunctionalTestEnvironment(environment) {
             stageWithAgent("Functional Test - AKS ${environment}", product) {
               testEnv(aksUrl) {
-                pcr.callAround("functionalTest:${environment}") {
-                  timeoutWithMsg(time: 40, unit: 'MINUTES', action: 'Functional Test - AKS') {
-                    builder.functionalTest()
+                pcr.callAround("functionalTest-aks:${environment}") {
```
Comment (timja) [inline vars/sectionDeployToAKS.groovy:89]: why?

File: vars/sectionPromoteBuildToStage.groovy [hunk 6]
```diff
@@ -39,7 +39,7 @@ def call(params) {
     withAcrClient(subscription) {
         def imageRegistry = env.TEAM_CONTAINER_REGISTRY ?: env.REGISTRY_NAME
         def projectBranch = new ProjectBranch(env.BRANCH_NAME)
-        def acr = new Acr(this, subscription, imageRegistry, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION)
+        def acr = new Acr(this, subscription, env.REGISTRY_NAME, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION, projectBranch)
```
Comment (timja) [inline vars/sectionPromoteBuildToStage.groovy:42]:
```suggestion
        def acr = new Acr(this, subscription, imageRegistry, env.REGISTRY_RESOURCE_GROUP, env.REGISTRY_SUBSCRIPTION, projectBranch)
```

File: vars/withAksClient.groovy [hunk 7]
```diff
@@ -2,7 +2,7 @@ import uk.gov.hmcts.contino.Environment
 
 def call(String subscription, String environment, String product, Closure block) {
   withAcrClient(subscription) {
-    def envName = environment.toUpperCase()
+    def envName = environment.replace('idam-', '').toUpperCase()
```
Comment (timja) [inline vars/withAksClient.groovy:5]:
```suggestion
    // replacing idam is a hack to workaround incorrect idam environment value
    def envName = environment.replace('idam-', '').toUpperCase()
```

File: vars/withParameterizedInfraPipeline.groovy [hunk 8]
```diff
@@ -6,10 +6,10 @@ import uk.gov.hmcts.contino.MetricsPublisher
 import uk.gov.hmcts.contino.Subscription
 import uk.gov.hmcts.pipeline.TeamConfig
 
-def call(String product, String environment, String subscription, Closure body) {
-  call(product, environment, subscription, '', body)
+def call(String product, String environment, String subscription, Boolean planOnly, Closure body) {
```
Comment (timja) [inline vars/withParameterizedInfraPipeline.groovy:9]: this is a breaking change and needs a new method without plan only added and passing false through in the `call` function

File: vars/withSubscriptionSp.groovy [hunk 9]
```diff
@@ -0,0 +1,74 @@
+#!groovy
+import groovy.json.JsonSlurperClassic
+
+def call(String subscription, Closure body) {
```
Comment (timja) [inline vars/withSubscriptionSp.groovy:4]: this doesn't appear to be called? can you explain this? possibly detach it to a separate PR as this is the only blocker from me right now

**Review by timja — CHANGES_REQUESTED** [review]

File: vars/sectionDeployToAKS.groovy (hunk 4 above)
Comment (nikola-naydenov-hmcts) [reply vars/sectionDeployToAKS.groovy:76]: I'm not sure, I will change it, I didn't actually write any of this

File: vars/sectionDeployToAKS.groovy (hunk 5 above)
Comment (nikola-naydenov-hmcts) [reply vars/sectionDeployToAKS.groovy:89]: 🤷‍♂️

File: vars/sectionDeployToAKS.groovy (hunk 4 above)
Comment (nikola-naydenov-hmcts) [reply vars/sectionDeployToAKS.groovy:76]: ah there was some closures in the Jenkins file     
```groovy
before('smoketest-aks:idam-preview') {
        env.PREVIEW_ENVIRONMENT_NAME = 'preview'
        env.NONPROD_ENVIRONMENT_NAME = 'preview'
        env.PUBLIC_URL = 'https://idam-web-public.service.core-compute-preview.internal'
        println """\
            Using PREVIEW_ENVIRONMENT_NAME: ${env.PREVIEW_ENVIRONMENT_NAME}
            Using NONPROD_ENVIRONMENT_NAME: ${env.NONPROD_ENVIRONMENT_NAME}
            Using PUBLIC_URL: ${env.PUBLIC_URL}""".stripIndent()
    }
```

Probably from when we where deploying to AKS and whatever the previous app container was called in different environments so we needed to differentiate.

File: vars/withParameterizedInfraPipeline.groovy (hunk 8 above)
Comment (nikola-naydenov-hmcts) [reply vars/withParameterizedInfraPipeline.groovy:12]: fair enough

File: vars/withSubscriptionSp.groovy (hunk 9 above)
Comment (nikola-naydenov-hmcts) [reply vars/withSubscriptionSp.groovy:4]: image

[Jenkinsfile_CNP](https://github.com/hmcts/cnp-idam-packer/blob/d43797286cdd8989d08c0f7b731e4de7ff6acb23/Jenkinsfile_CNP)
[Jenkinsfile_parameterized](https://github.com/hmcts/cnp-idam-packer/blob/d43797286cdd8989d08c0f7b731e4de7ff6acb23/Jenkinsfile_parameterized)
[README.md](https://github.com/hmcts/cnp-idam-master/blob/cbd38b355f53620d61bba24de64abef70da769a6/README.md)
It is being used, but .. no I can't explain it, Henry Dobson and Douglas Brand Williamson implemented it. 
It just seems like they need a specific set of variables set to do.. something..

Comment (nikola-naydenov-hmcts) [reply vars/withSubscriptionSp.groovy:4]: if I compare it with the standard **_withSubscription_** the difference is this:
image
 you always call 
```groovy
        servicePrincipalBasedLogin(subscription, body)
```
its otherwise identical

Comment (timja) [reply vars/withSubscriptionSp.groovy:4]: I think that was because packer didn't support managed identity, which looks like it might right now, https://www.packer.io/docs/builders/azure#azure-managed-identity

Again can we detach to this to a different PR? As I think this shouldn't be needed now, support for this was intentionally removed.

Comment (nikola-naydenov-hmcts) [reply vars/withSubscriptionSp.groovy:4]: sure, I will try re-introduce it if it turns out we cannot manage without it

**Review by timja — CHANGES_REQUESTED** [review]
unit tests are currently failing

**Comment by nikola-naydenov-hmcts** [comment]
> unit tests are currently failing

I fixed the tests by changing this:

```groovy
    def buildTag = projectBranch.isPR() ? dockerImage.getBaseTaggedName() : dockerImage.getTaggedName()
    def buildTag = projectBranch.isPR() ? dockerImage.getBaseTaggedName() : dockerImage.getBaseShortName()
```

Gotta be honest though, I don't really understand why we have it like that in the Branch as in current master its always getBaseShortName, and the methods seem pretty interchangeable

File: src/uk/gov/hmcts/contino/azure/Acr.groovy [hunk 10]
```diff
@@ -76,7 +79,8 @@ class Acr extends Az {
    *   stdout of the step
    */
   def build(DockerImage dockerImage, String additionalArgs) {
-    this.az "acr build --no-format -r ${registryName} -t ${dockerImage.getBaseShortName()} --subscription ${registrySubscription} -g ${resourceGroup} --build-arg REGISTRY_NAME=${registryName}${additionalArgs} ."
+    def buildTag = projectBranch.isPR() ? dockerImage.getBaseTaggedName() : dockerImage.getBaseShortName()
```
Comment (timja) [inline src/uk/gov/hmcts/contino/azure/Acr.groovy:82]: why this change? it seems like this uses the full registry name on a PR?

Comment (nikola-naydenov-hmcts) [reply src/uk/gov/hmcts/contino/azure/Acr.groovy:82]: I.. honestly have no idea, most of the changes are for the purposes of building on preview as an actual branch..
would this help or affect that adversely? 
I don't see why it would

Comment (timja) [reply src/uk/gov/hmcts/contino/azure/Acr.groovy:82]: exactly, this is my last concern atm, could you try reverting it and we can see if it works for you?

Comment (nikola-naydenov-hmcts) [reply src/uk/gov/hmcts/contino/azure/Acr.groovy]: Okay I have reverted that change, test seem to be passing.
When I try deploy the web-apps to preview from master, we will see if that did anything before 🤔

File: test/uk/gov/hmcts/contino/azure/AcrTest.groovy [hunk 11]
```diff
@@ -15,10 +16,12 @@ class AcrTest extends Specification {
   def steps
   def acr
   def dockerImage
+  def projectBranch
 
   def setup() {
     steps = Mock(JenkinsStepMock.class)
     dockerImage = Mock(DockerImage.class)
+    projectBranch = new ProjectBranch('test')
```
Comment (timja) [inline test/uk/gov/hmcts/contino/azure/AcrTest.groovy:24]: looks unneeded now?

File: test/uk/gov/hmcts/contino/azure/AcrTest.groovy [hunk 12]
```diff
@@ -15,10 +16,12 @@ class AcrTest extends Specification {
   def steps
   def acr
   def dockerImage
+  def projectBranch
```
Comment (timja) [inline test/uk/gov/hmcts/contino/azure/AcrTest.groovy:19]: ?

File: test/uk/gov/hmcts/contino/HelmTest.groovy [hunk 13]
```diff
@@ -17,7 +17,8 @@ class HelmTest extends Specification {
     steps = Mock(JenkinsStepMock.class)
     steps.env >> [AKS_RESOURCE_GROUP: "cnp-aks-rg",
                   AKS_CLUSTER_NAME: "cnp-aks-cluster",
-                  SUBSCRIPTION_NAME: "${SUBSCRIPTION}"]
+                  SUBSCRIPTION_NAME: "${SUBSCRIPTION}",
+                  BRANCH_NAME: "master"]
```
Comment (timja) [inline test/uk/gov/hmcts/contino/HelmTest.groovy:21]: looks like this change can be reverted?

File: test/uk/gov/hmcts/contino/azure/AcrTest.groovy (hunk 11 above)
Comment (nikola-naydenov-hmcts) [reply test/uk/gov/hmcts/contino/azure/AcrTest.groovy]: right

File: test/uk/gov/hmcts/contino/HelmTest.groovy (hunk 13 above)
Comment (nikola-naydenov-hmcts) [reply test/uk/gov/hmcts/contino/HelmTest.groovy]: true

File: test/uk/gov/hmcts/contino/azure/AcrTest.groovy (hunk 12 above)
Comment (nikola-naydenov-hmcts) [reply test/uk/gov/hmcts/contino/azure/AcrTest.groovy]: fair enough

**Review by timja — APPROVED** [review]

**Comment by nikola-naydenov-hmcts** [comment]
Can someone merge this for me, I am unauthorized
